sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -385,11 +396,16 @@ func (n *DataSource) Checkpoint() (SplitResult, time.Duration, bool, error) {
 	if err != nil {
 		return SplitResult{}, -1 * time.Minute, false, err
 	}
+	if len(rs) == 0 {
+		return SplitResult{}, -1 * time.Minute, false, nil
```

Comment:
I'm trying to wrap my head around when we would ever expect this case - I have 2 related questions:

1) If the user has checkpointed but then returns an empty residual, they shouldn't have checkpointed, right? I'd expect us to at least warn in that case probably.
2) Even if there are no residuals, don't we still want to validate that they haven't set any primaries? That's still an error waiting to happen